pkg/schedule/affinity/policy.go

```diff
@@ -0,0 +1,232 @@
+	if !m.IsAvailable() {
+		return
+	}
+	unavailableStores := m.generateUnavailableStores()
+	isUnavailableStoresChanged, groupAvailabilityChanges := m.getGroupAvailabilityChanges(unavailableStores)
+	if isUnavailableStoresChanged {
+		m.setGroupAvailabilityChanges(unavailableStores, groupAvailabilityChanges)
+	}
+	m.collectMetrics()
+}
+
+// collectMetrics collects the global metrics of the affinity manager.
+func (m *Manager) collectMetrics() {
+	m.RLock()
+	defer m.RUnlock()
+
+	// Collect global metrics
+	groupCount.Set(float64(len(m.groups)))
+	regionCount.Set(float64(len(m.regions)))
+	affinityRegionCount.Set(float64(m.affinityRegionCount))
+}
+
+func (m *Manager) generateUnavailableStores() map[uint64]storeCondition {
+	unavailableStores := make(map[uint64]storeCondition)
+	stores := m.storeSetInformer.GetStores()
+	lowSpaceRatio := m.conf.GetLowSpaceRatio()
+	for _, store := range stores {
+		switch {
+		// First the conditions that will mark the group as expired
+		case store.IsRemoved() || store.IsPhysicallyDestroyed() || store.IsRemoving():
+			unavailableStores[store.GetID()] = storeRemovingOrRemoved
+		case store.IsUnhealthy():
+			unavailableStores[store.GetID()] = storeDown
+
+		// Then the conditions that will mark the group as degraded
+		case !store.AllowLeaderTransferIn() || m.conf.CheckLabelProperty(config.RejectLeader, store.GetLabels()):
+			unavailableStores[store.GetID()] = storeEvictLeader
+		case store.IsDisconnected():
+			unavailableStores[store.GetID()] = storeDisconnected
+		case store.IsLowSpace(lowSpaceRatio):
+			unavailableStores[store.GetID()] = storeLowSpace
+		case store.IsPreparing():
+			unavailableStores[store.GetID()] = storePreparing
+		}
+		// Note: We intentionally do NOT check:
+		// - IsSlow(): Performance issue, not availability issue
+	}
+	return unavailableStores
+}
+
+func (m *Manager) getGroupAvailabilityChanges(unavailableStores map[uint64]storeCondition) (isUnavailableStoresChanged bool, groupAvailabilityChanges map[string]groupAvailability) {
+	var logEntries []logEntry
+	groupAvailabilityChanges = make(map[string]groupAvailability)
+
+	// Validate whether unavailableStores has changed.
+	m.RLock()
+	isUnavailableStoresChanged = !maps.Equal(unavailableStores, m.unavailableStores)
+	if !isUnavailableStoresChanged {
+		m.RUnlock()
+		return false, nil
+	}
+
+	// Analyze which Groups have changed availability
+	// Collect log messages to print after releasing lock
+	for _, groupInfo := range m.groups {
+		var unavailableStore uint64
+		var maxCondition storeCondition
+		for _, storeID := range groupInfo.VoterStoreIDs {
+			if condition, ok := unavailableStores[storeID]; ok && (!condition.affectsLeaderOnly() || storeID == groupInfo.LeaderStoreID) {
+				if unavailableStore == 0 || condition > maxCondition {
+					unavailableStore = storeID
+					maxCondition = condition
+				}
+			}
+		}
+		newAvailability := maxCondition.groupAvailability()
+		if newAvailability != groupInfo.GetAvailability() {
```

Comment:
Please add some comments to explain why when availability changing, `unavailableStore != 0` represents that group has a new unavailableStore, otherwise represents group becomes health.

This contains a great deal of underlying logical relationships. Perhaps modifying the code could make this section simpler and easier to understand.

Reply:
refactor it